### DISCUSSION TITLE
Shapes, SVG and Ricard Marxer's Geomerative

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#473] [Q&A]
When I was working with Processing I counted on [Geomerative](http://www.ricardmarxer.com/geomerative/) (the site is not loading for me, but there is this [repo](https://github.com/rikrd/geomerative)) for a lot of computational geometry stuff that I now do with [shapely](https://shapely.readthedocs.io/). 

Back then, before I knew about the undocumented `PFont.get_shape()` for glyph shape extraction, I had to use Geomerative to "distort letters" inspired by [André Burnier](https://garoa.net.br/wiki/Noite_de_Processing#29.2F09.2F2020:_.5BONLINE.5D_10_maneirad_de_distorcer_letras_com_Processing_-_Andr.C3.A9_Burnier), also, it had good SVG loading support if I remember correctly. 

There are some things that I still have to find out how to do... Geomerative had a [polygonizer feature](https://github.com/rikrd/geomerative/blob/master/examples/Tutorial_10_HelloPolygonize/Tutorial_10_HelloPolygonize.pde) that would simplify a shape, I think it would be useful to get bezier/cubic based SVGs and make a `shapely.Polygon` from them. I might have a look at its source maybe...

Why would I want those polygons? Because then I can "triangulate" them which is useful to simulate concave shapes in [pymunk's 2D physics engine](http://www.pymunk.org/en/latest/)...

**Comment by hx2A** [comment]
Processing (and py5) can load an SVG file, and then you can get the vertices from the Py5Shape object. Does that work?

**Reply by villares** [reply]
It kind of works, but dealing with the different types of curve is really a chore I'd like to avoid. I just want polygons approximating the curves. One would have to check [`.get_vertex_code()`](http://py5coding.org/reference/py5shape_get_vertex_code.html) as sometimes `.get_vertex()` will be a control point, etc.

**Reply by villares** [reply]
Last week I had some strange behavior with a student: comparing the geometry I got from `shp_path_child.get_vertex()` on a Py5Shape from a loaded SVG, it was kind of scaled/shrunk when compared with the result from `shape(shp, 0, 0)` ,  ... I'll try to reproduce on Processing Java and py5 when I get the chance.

**Reply by villares** [reply]
There is some tranformation going on  (close to a 3.5 scaling) happening between the path and the full shape loaded from the SVG...

![image](https://github.com/py5coding/py5generator/assets/3694604/2d110099-21bf-4264-8e12-80505235c5f5)

**Reply by hx2A** [reply]
I'm not sure what is going on here, I'd have to look at the Processing source code. There might be a "scale" operation in a parent shape that is getting omitted when you attempt to draw a child shape directly, without going through the parent shape.

You might get different results if you use the P2D or P3D renderers. Can you give that a try?

**Reply by villares** [reply]
Cool guess huh! P2D and P3D both act the same, but differently from the default renderer!
![image](https://github.com/py5coding/py5generator/assets/3694604/6f891f82-c297-410d-99aa-de71fb2ed21c)

**Reply by villares** [reply]
Same on py5, if you want to play with it.

```python
import py5

def setup():
    global shp
    py5.size(300, 300, py5.P2D)
    shp = py5.load_shape('a.svg')
    path = shp.get_children()[1].get_children()[0]
    py5.translate(50, 100)
    py5.shape(shp, 0,0)
    py5.shape(path, 0, 0)
    py5.shape(shp, 100, 0)
    py5.shape(shp, 0, 0)
    py5.scale(3.5)
    path.set_stroke(255)
    path.set_fill(py5.color(255, 0, 0, 10))
    py5.shape(path, 0, 0)
    py5.save('out.png')
    

py5.run_sketch(block=False)
```

And the SVG:
![a](https://github.com/py5coding/py5generator/assets/3694604/7f78c280-39e7-4d5d-988e-3bc14c7a23d6)

**Reply by hx2A** [reply]
> Cool guess huh! P2D and P3D both act the same, but differently from the default renderer!

Yes, I figured that would be the case! The P2D and P3D renderers are similar to each other in that they use tessellated geometry internally. I default renderer does very different things for shapes.

**Reply by hx2A** [reply]
@villares I just came across this library:

https://github.com/regebro/svg.pathhttps://github.com/regebro/svg.path

It might be useful to you and to py5. I don't have time to do a deep dive into this but the first glance looked very promising.

If you think this is useful, I'll investigate it as a new py5 integration.

**Reply by villares** [reply]
Very nice! This other one also caught my eye: https://github.com/mathandy/svgpathtools
Ha:
> Much of the core of this module was taken from [the svg.path (v2.0) module](https://github.com/regebro/svg.path). Interested svg.path users should see the compatibility notes at bottom of this readme.

**Reply by hx2A** [reply]
Yes, SVG Path tools is also useful! Another candidate for integration with py5.
